exec credential provider: InteractiveMode support

ankeesler opened the pull request:
#### What type of PR is this?

/kind bug

(My guilty conscious is saying that this is really actually a feature, but I am submitting it as a bug because we would like to fix https://github.com/kubernetes/kubernetes/issues/98451 ASAP)

#### What this PR does / why we need it:

- Allow exec plugin authors to control how client-go treats standard input with respect to an exec plugin; this allows for more intelligent handling of standard input
- Add `users.user.exec.interactiveMode` to kubeconfig which is used to specify how standard input should be handled in relation to the exec plugin
- Add `ExecCredential.Spec.Interactive` to communicate to exec plugin whether standard input was passed to it
- `exec.GetAuthenticator()` fails if the exec plugin requires standard input, but it is not available

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/98451

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Exec plugin authors can override default handling of standard input via new interactiveMode kubeconfig field
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP tracking: https://github.com/kubernetes/enhancements/issues/541

/sig auth
/cc @liggitt @enj

FYI - I have been using https://github.com/ankeesler/sample-exec-plugin to test this manually.

ankeesler commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types.go:
```diff
@@ -46,6 +46,9 @@ type ExecCredentialSpec struct {
 	Cluster *Cluster `json:"cluster,omitempty"`
+
+	// Interactive declares whether stdin has been passed to this exec plugin.
+	Interactive bool `json:"interactive"`
```
Open question: do we like this contract with the exec plugin?

Some discussion here: https://kubernetes.slack.com/archives/C0EN96KUY/p1613768159114200?thread_ts=1613767745.112500&cid=C0EN96KUY.

ankeesler commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -212,6 +217,27 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
 	return c.put(key, a), nil
 }
 
+func isInteractive(isTerminalFunc func(int) bool, config *api.ExecConfig) (bool, error) {
```
Open question: how do we ensure that a higher level entity is not using standard input?

I think this is hard to do. Some discussion here:
- https://github.com/kubernetes/kubernetes/issues/98451#issuecomment-783501007
- https://kubernetes.slack.com/archives/C0EN96KUY/p1613768319115600?thread_ts=1613767745.112500&cid=C0EN96KUY

enj replied:
I am wondering if could probe `os.Stdin` somehow to tell if `kubectl` is already using it for something else (in `UpdateTransportConfig` or `getUserIdentificationPartialConfig` or ...).

TODO: look into this more next week.

ankeesler replied:
I also have been looking into similar credential plugins to see if there is any prior art here. TLDR: I can't find anyone else that is hooking up a TTY to stdin on the credential plugin.
- Docker credential plugins (takes command input on stdin): https://docs.docker.com/engine/reference/commandline/login/#credential-helpers
- Git credential plugins (takes command input on stdin): https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage
- (Other? I went through all of the CLI tools that I used to connect to infrastructure, but didn't find any good examples)

ankeesler replied:
Redirecting this conversation to https://github.com/kubernetes/kubernetes/pull/99310#discussion_r613462261.

ankeesler commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -181,6 +181,11 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
 		connTracker,
 	)
 
+	interactive, err := isInteractive(isTerminalFunc, config)
```
Do we want to move this check to take place lazily in `refreshCredsLocked()`? I wonder if someone could hook up a TTY to standard input on a long-running client-go process. Regardless, this seems like an anti-pattern, so I only performed the check once here.

enj replied:
Checking it once seems fine to me, though maybe we should check it in `UpdateTransportConfig`?

ankeesler replied:
Mmm, I think I just thought of a reason we want this check down in `refreshCredsLocked()`.
1. User types `kubectl create -f -`
2. The kubectl code wires up a `kubernetes.Clientset` which transitively calls `GetAuthenticator()` and `UpdateTransportConfig()` (note: `terminal.IsTerminal()` returns `true` here since standard input is open and a terminal)
3. The kubectl code goes to get the object from standard input
4. User types in their object by hand on the command line
5. User closes standard input with Ctrl-D
6. kubectl receives object from standard input
7. kubectl makes API call with client
8. client goes to refresh exec credentials (note: `terminal.IsTerminal()` returns `false` here since standard input is closed)

liggitt replied:
maybe... but I still think we want something in place that lets `kubectl -f -` mark stdin as unavailable for use by credential plugins to avoid that entirely

I *think* the resource visitor could handle stdin containing two objects:

```
kind: ...
apiVersion: ....
...
---
kind: ...
apiVersion: ....
...
```

by reading the first, making an API call, then reading the second and making another API call

we wouldn't want the credential plugin to consume the second object as stdin

ankeesler replied:
mmm. yeah. that makes sense.

here are the current constraints/design goals/opinions that i can think of for this issue:
- applications built on top of client-go should tell client-go whether stdin is available for use by credential plugins
- we should pick a safe default for this behavior so that when an existing application upgrades to client-go v0.22.0 AND that existing application consumes a kubeconfig with a client.authentication.k8s.io/v1 interactive exec plugin, then the interactive behavior will work safely
- it would be really nice if that _safe default for this behavior_ from above would involve client.authentication.k8s.io/v1 interactive exec plugins working out of the box with the existing client-go application that has just upgraded to client-go v0.22.0, since commonly these two components (client-go app and exec plugin) will be developed and distributed by completely separate organizations
- all client-go applications that consume a kubeconfig should care about whether stdin is available for use, because they do not necessarily know what underlying authentication mechanism is in the kubeconfig

we can somewhat easily solve the backend of this flow by adding a new field/parameter to rest config creation that determines whether stdin is safe to use (or we can provide the stdin `os.File` itself?). but i haven't envisioned how we would wire this down from the top of the stack :(

ankeesler replied:
> but I still think we want something in place that lets kubectl -f - mark stdin as unavailable for use by credential plugins to avoid that entirely

Alright, I gave this a shot.
- client-go changes: https://github.com/kubernetes/kubernetes/pull/99310/commits/ad7a50bddefcdc1b9f41bc17e38e1c4f3fc45772
- `kubectl` changes: https://github.com/kubernetes/kubernetes/pull/99310/commits/6070cff07e8a3137017e3527e92ee64e80aa5164

What do y'all think? Do we think this is good enough to protect against the "stdin fighting" problem?

liggitt replied:
current approach looks reasonable to me

ankeesler commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -212,6 +217,27 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+
+	switch *config.InteractiveMode {
+	case api.NeverExecInteractiveMode:
+		return false, nil
```
I checked and I don't think we can do set `stdin` to an `io.Reader` that detects if a `Read()` took place in order to provide the user with a nice error message. This is because the go standard library tried to read that `io.Reader` in a loop and so it will _always_ get `Read()`, even if the exec plugin on the other end of the pipe does not `Read()`.

https://github.com/golang/go/blob/0f66fb7b856b02497cf801ce72d80f375f53358b/src/os/exec/exec.go#L266

enj replied:
meh, that is a bummer if true.

TODO: look into this more next week.

enj replied:
I do see some special handling of `*os.File`, maybe we could make one using `os.NewFile` and try to see if that got read?

ankeesler replied:
I made a helper program to try this out. The helper program is really messy, but it should run and teach us something. https://gist.github.com/ankeesler/7173d29d4ac81c8b97b6c584836db142

Here is what I learned:
- If a parent process opens a pipe and passes it to a child process, it can read from that pipe to determine whether the child process has read from the pipe (I think I convinced myself of this since the child process gets an exact copy of the parent's file descriptor table, so that same pipe file descriptor will have the same seek offset for both processes)

So my next question is...if the parent process writes something to a pipe and then passes the read end to the child process, then how can the child process tell the difference between something typed from the user and a sentinel value from the parent process?

For example, if the parent process was to write a single byte to the write end of the pipe, and then the child process read from the pipe, then the child process would get a single byte, which they could suspect came from the user.

ankeesler replied:
Correct me if I am wrong, but on Zoom we discussed punting on this for now.

enj commented:
/ok-to-test

fedebongio commented:
/remove-sig api-machinery

enj commented:
/kind feature
/kind api-change
/priority important-longterm
/triage accepted

enj reviewed (changes requested):
First pass review.  Have some TODOs for me for next week.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -212,6 +217,27 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+func isInteractive(isTerminalFunc func(int) bool, config *api.ExecConfig) (bool, error) {
+	if config.InteractiveMode == nil {
+		// backwards compatible default
+		return isTerminalFunc(int(os.Stdout.Fd())), nil
```
It's debatable if we want preserve this (incorrect?) behavior.

ankeesler replied:
Did we decide over Zoom that we would need to preserve this (incorrect) behavior for the sake of v1beta1 backwards compatibility? For v1 I think we decided we want to use `os.Stdin.Fd()` only.

ankeesler replied:
Fixed this to match 1.21 bug fix behavior: https://github.com/kubernetes/kubernetes/pull/99310/commits/4acb159638e01eb707bf8247dd6e88bcf2f29969.

liggitt replied:
resolved by IfAvailable approach which matches old behavior and is a generally useful thing for a plugin to be able to express

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/types.go:
```diff
@@ -245,6 +245,10 @@ type ExecConfig struct {
+
+	// InteractiveMode determines this plugin's relationship with standard input.
+	// +optional
+	InteractiveMode *ExecInteractiveMode `json:"interactiveMode,omitempty"`
```
458d12b02bef149a10552d172f5d3762e36b1e2e

ankeesler replied:
6c67484945d5df901006bef98960219dbd9ca8a5

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/types.go:
```diff
@@ -245,6 +245,10 @@ type ExecConfig struct {
+
+	// InteractiveMode determines this plugin's relationship with standard input.
+	// +optional
+	InteractiveMode *ExecInteractiveMode `json:"interactiveMode,omitempty"`
```
458d12b02bef149a10552d172f5d3762e36b1e2e

ankeesler commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types.go:
```diff
@@ -46,6 +46,9 @@ type ExecCredentialSpec struct {
 	Cluster *Cluster `json:"cluster,omitempty"`
+
+	// Interactive declares whether stdin has been passed to this exec plugin.
+	Interactive bool `json:"interactive"`
```
To play it safe, I changed the name to `IsInteractive` as a next pass. I'm not totally in love with it.

f22b295250e1426432cc32652e0a28c4c0a00a23

liggitt replied:
why wouldn't we reuse the field name... (and back-fill it to the beta version as well)? it's the same meaning, right?

ankeesler replied:
i changed it back to `Interactive` and updated the comment that @enj referenced. the fields are logically the same (they are set to the same value) and i think they are semantically the same as well (they mean "this plugin can use standard input to interact with the user").

daa7936743684065176a0c596a54afeee57d6ed2

ankeesler commented:
marking this as a WIP PR, since we are probably gonna put this through API review before we come back to the implementation in the 1.22 timeframe

ankeesler commented on staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:
```diff
@@ -365,6 +366,7 @@ func (b *Builder) URL(httpAttemptCount int, urls ...*url.URL) *Builder {
 // will be ignored (but logged at V(2)).
 func (b *Builder) Stdin() *Builder {
 	b.stream = true
+	b.stdin = true
```
Note! There are other places where `kubectl` can potentially use standard input. I have not added code to protect against stdin fighting in these places because `kubectl` uses standard input once and then never looks at it again. Thus, `kubectl` would never try to use standard input when the exec plugin might be trying to use it.
- https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L283-L290
- https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/kubectl/pkg/rawhttp/raw.go#L58
- https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/kubectl/pkg/cmd/set/env/env_parse.go#L68
- https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/kubectl/pkg/cmd/replace/replace.go#L310
- Are there more????

How do we protect against someone adding more `kubectl` code that reads from stdin and may cause a stdin fight between `kubectl` and an exec plugin? I have no idea how to future-proof this???

enj replied:
I do not know of a good way to keep this working perfectly "forever."

liggitt replied:
- https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/kubectl/pkg/cmd/set/env/env_parse.go#L68
- https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/kubectl/pkg/cmd/replace/replace.go#L310


How can we hook these two uses up to the StdinUnavailable bits?

ankeesler replied:
`env_parse.go`: we could return a boolean from `envutil.ParseEnv()` for whether stdin was used in, and then wire that into the resource builder here (we would need a new `resource.Builder` function, which seems kinda leaky): https://github.com/kubernetes/kubernetes/blob/74497632f4118e7989da73de0ed6f1268aeb64ad/staging/src/k8s.io/kubectl/pkg/cmd/set/set_env.go#L273

`replace.go`: if we did add a new `resource.Builder` function, we could call it again here; again, this seems a little leaky to me

ankeesler replied:
i added a number of commits to try to address @liggitt’s comments:
* 1a2957cebab: updated `resource.Builder` to contain function to say when stdin is in use (seems leaky, but best i could do)
* b5c4c593664: marked stdin as in use in `kubectl set env` (note - i ran into https://github.com/kubernetes/kubernetes/issues/102741 when trying to test this)
* f8599f88f9e: added cmd test for when someone tries to use stdin for both the resource and env vars in `kubectl set env`; because of these new `resource.Builder` changes, users will see a new error on the CLI that will prevent them from shooting themselves in the foot:
```akeesler@akeesler-a02:kubernetes$ echo a=b | ./_output/local/bin/darwin/amd64/kubectl set env -f - -
error: standard input cannot be used for multiple arguments```
* e5af8c48475: marked stdin as in use in `kubectl replace --force`
* 441ab1784f3: added cmd tests for when these commands are used with `Always` interactive exec plugin

liggitt replied:
those changes look great, thanks. one nit on https://github.com/kubernetes/kubernetes/commit/b5c4c5936642e6f05ded12d60cfb2f9775cba8b2#r52149752 to track/return whether stdin was consumed even in error cases

ankeesler replied:
nit addressed in https://github.com/kubernetes/kubernetes/pull/99310/commits/20193d72149e81a410334e3c4145ab57b9b40687

sftim commented:
If the exec plugin can read from stderr, is that a viable option?

https://github.com/jleffler/soq/tree/master/src/so-4844-3136 explains how a C program can treat stderr as an input. Useful to avoid swallowing input from stdin when you want to prompt for credentials.

ankeesler commented:
@sftim - good thinking. The same thing works in go with `os.Stderr`. Right now the exec plugin contract states that output (e.g., debug info, prompting the user, etc.) should go to stderr, but since `os.File`'s (`FILE*`'s in the C example) are full duplex, then the streams shouldn't collide.

This begs the question...even if an exec plugin reads from stderr, how would it differentiate from what is on stdin? They are essentially the same "stream" (i.e., TTY), right?

sftim commented:
> even if an exec plugin reads from stderr, how would it differentiate from what is on stdin? They are essentially the same "stream" (i.e., TTY), right? 

The usual logic is that if stderr and stdin are in fact the two different file descriptors referencing the same TTY, read from stdin. If stdin is not a TTY and stderr is a TTY, and you need a terminal to prompt for input, write the prompt to stderr and read from stderr.

Related: you can also exec a tool like pinentry (see https://github.com/foxcpp/go-assuan/tree/master/pinentry for a related Go project) and make its stdout and stdin be nothing to do with the user's TTY, but pass it your stderr unchanged.

ankeesler commented:
> stderr and stdin are in fact the two different file descriptors referencing the same TTY

Yeah, I think I phrased my previous comment incorrectly. From the user's perspective, they can't differentiate between stderr and stdin, right? They only "see" the one TTY? So if they entered in text in their terminal, then they could still potentially get confused as to which process (parent or child) that the text would be sent to?

> Related: you can also exec a tool like pinentry

Agreed that this is probably a good practice for some exec plugins. I still feel like we should allow exec plugins should be able to simply read from a standard stream so that they aren't forced to integrate a tool (e.g., pinentry) into their project.

sftim commented:
I guess my quibble is really about:
> Add `ExecCredential.Spec.Interactive` to communicate to exec plugin whether standard input was passed to it

The docs say:
> When run from an interactive session, stdin is exposed directly to the plugin. Plugins should use a TTY check to determine if it's appropriate to prompt a user interactively.

I'm thinking that we might clarify that if `.spec.interactive` is true in `KUBERNETES_EXEC_INFO` then kubectl can choose what TTY-like file descriptor it passes. From the exec credential plugin's point of view this is stdin, but we don't guarantee that this is also the stdin of kubectl. Useful if you wanted to `kubectl apply -f -` but still provide credentials.

(I'm mentioning this now because this kind of thing is hard to redefine later).

sftim commented:
:thought_balloon: Can I also specify `interactiveMode: false` in the KUBECONFIG to tell kubectl to pass /dev/null as stdin?

ankeesler commented:
> The docs say:

Ah, thanks for reminding me about docs. We definitely want to make changes to those after this feature. I added docs stuff to my unofficial TODO list here: https://github.com/kubernetes/enhancements/issues/541#issuecomment-799372909.

> kind of thing is hard to redefine later

Yeah, I think it would be valuable to make the contract looser to allow for future flexibility.

> 💭 Can I also specify interactiveMode: false in the KUBECONFIG to tell kubectl to pass /dev/null as stdin?

Yes - more specifically, you can say `interactiveMode: Never` and `kubectl` would always just pass `/dev/null` as stdin to the exec plugin.

enj commented on staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:
```diff
@@ -74,6 +74,7 @@ type Builder struct {
 
 	paths  []Visitor
 	stream bool
+	stdin  bool
```
Unclear what this means when looking at it out of context.

```suggestion
	stdinInUse  bool
```

enj commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:
```diff
@@ -357,6 +360,18 @@ func cleanANSIEscapeCodes(s string) string {
 
+func validateInteractiveMode(mode clientcmdapi.ExecInteractiveMode) error {
+	switch mode {
+	case "":
+		// Empty mode is fine, for now.
+		return nil
```
I think we can add defaulting to the beta API (but not v1) that would set this as `IfAvailableExecInteractiveMode` and have the rest of the code assume it is always set.  So the empty string case here would return a validation error similar to `field.Required`.

ankeesler replied:
Since this `InteractiveMode` field would be on the v1 _clientcmd_ API, would we write a defaulting func that looks at the `ExecConfig.APIVersion` to default the `ExecConfig.ExecInteractiveMode`? I tried that here: 54dd5407693621307d2dc61a9bcc2c3e832d472e.

enj commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -89,6 +89,12 @@ type Config struct {
+	// input through to the exec plugin. For example, a higher level entity might be using
+	// standard input for something else and therefore it would not be safe for the exec
+	// plugin to use standard input.
+	ExecStdinUnavailable bool
```
Should this be inside `ExecConfig`?  Similar to `LocationOfOrigin`?

```go
// Cluster contains information about how to communicate with a kubernetes cluster
type Cluster struct {
	// LocationOfOrigin indicates where this object came from.  It is used for round tripping config post-merge, but never serialized.
	// +k8s:conversion-gen=false
	LocationOfOrigin string
...
}

ankeesler replied:
I'm glad you brought this up. I didn't put it in `ExecConfig` because I didn't think we wanted a user to be able to set it in their kubeconfig. But I do agree this is weird to have it hanging off `rest.Config`. It looks like `LocationOfOrigin` follows the pattern of "this field should logically be on the kubeconfig struct, but it shouldn't be serialized".

Gave this a go here: e91ed5ce232cdd45d0e3bc719e0e5ee3454601bb.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -213,6 +215,32 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+func isInteractive(isTerminalFunc func(int) bool, config *api.ExecConfig, stdinUnavailable bool) (bool, error) {
+	var shouldBeInteractive bool
+	switch config.InteractiveMode {
+	case "":
+		// backwards compatible default
+		shouldBeInteractive = isTerminalFunc(int(os.Stdin.Fd()))
```
Defaulting should allow you to drop this case.

ankeesler replied:
If folks are manually setting their `rest.Config.ExecConfig` today, then they will most likely run into a runtime error here upon upgrade with no changes to their code. Is this OK? Per some of our previous conversations about upgrading k8s.io Go code, it sounds like there is precedent for this.

Similar case for `clientcmd.Config` - folks that are setting their `clientcmd.Config` manually will most likely get a validation error on upgrade if they are using an `ExecConfig` and they don't set the `InteractiveMode` field.

See 54dd5407693621307d2dc61a9bcc2c3e832d472e for where I took a pass at defaulting.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -213,6 +215,32 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+		return false, fmt.Errorf("unknown ExecInteractiveMode: %q", config.InteractiveMode)
+	}
+
+	if shouldBeInteractive && stdinUnavailable {
+		return false, fmt.Errorf("standard input is unavailable")
+	}
```
Is this correct for when `mode=IfAvailableExecInteractiveMode`, `shouldBeInteractive=true` and `stdinUnavailable=true`?

ankeesler replied:
This should now be correct, and read a little bit better, for when `mode=IfAvailableExecInteractiveMode` and `isTerminalFunc()` returns `true` thanks to 76281c20ccd.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec_test.go:
```diff
@@ -517,7 +522,176 @@ func TestRefreshCreds(t *testing.T) {
+			}`,
+			wantCreds: credentials{token: "foo-bar"},
+		},
+		{
+			name: "beta-basic-request-with-if-available-interactive-mode-and-terminal-and-stdin-unavailable",
+			config: api.ExecConfig{
+				APIVersion:      "client.authentication.k8s.io/v1beta1",
+				InteractiveMode: api.IfAvailableExecInteractiveMode,
+			},
+			stdinUnavailable: true,
+			isTerminal:       true,
+			wantErr:          true,
+			wantErrSubstr:    "exec plugin cannot support interactive mode: standard input is unavailable",
+		},
```
This should run without error right and just set `"isInteractive": false`?

ankeesler replied:
Dang. You're right. I missed this. 76281c20ccd.

enj reviewed (changes requested):
I do not know how we make sure that `kubectl` does not consume stdin without us noticing.

ankeesler commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -375,10 +397,16 @@ func (a *Authenticator) maybeRefreshCreds(creds *credentials, r *clientauthentic
-			Response:    r,
-			Interactive: a.interactive,
+			Response:      r,
+			Interactive:   interactive,
+			IsInteractive: interactive,
```
Maybe we can riff on this in https://github.com/kubernetes/kubernetes/pull/99310#discussion_r623363566.

liggitt replied:
commented there

ankeesler commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/zz_generated.conversion.go:
```diff
@@ -119,6 +119,7 @@ func Convert_v1alpha1_ExecCredentialSpec_To_clientauthentication_ExecCredentialS
 func autoConvert_clientauthentication_ExecCredentialSpec_To_v1alpha1_ExecCredentialSpec(in *clientauthentication.ExecCredentialSpec, out *ExecCredentialSpec, s conversion.Scope) error {
 	out.Response = (*Response)(unsafe.Pointer(in.Response))
 	out.Interactive = in.Interactive
```
I started with `Interactive`, and then @enj made this point: https://github.com/kubernetes/kubernetes/pull/99310#discussion_r583876986. It does seem nicer to use the same field name, if we can get away with it.

liggitt replied:
I think we should use the same name

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/v1/defaults.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+// SetDefaults_Config sets defaults for a Config
+func SetDefaults_Config(config *Config) {
```
Whoa...magic. I messed around with this generated defaulting thing for a while yesterday but couldn't get it working. 279b152f07a.

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go:
```diff
@@ -229,3 +233,22 @@ type ExecEnvVar struct {
+	IfAvailableExecInteractiveMode ExecInteractiveMode = "IfAvailable"
+	// AlwaysExecInteractiveMode declares that this exec plugin requires standard input in order to
+	// run, and therefore the exec plugin will only be run if standard input is available for user
+	// input.
```
Gave this another shot in 88fa67fc956.

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go:
```diff
@@ -221,6 +221,10 @@ type ExecConfig struct {
 	// reading this environment variable.
 	ProvideClusterInfo bool `json:"provideClusterInfo"`
+
+	// InteractiveMode determines this plugin's relationship with standard input.
```
describe the valid values, the default behaviors for v1alpha1 and v1beta1, and the requirement to specify it for v1

ankeesler replied:
9bd7adeb5b3ba24ee2b10eab7c15b33beb009afb

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -213,6 +213,28 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+			return false, errors.New("standard input is not a terminal")
+		}
+		if config.StdinUnavailable {
+			return false, fmt.Errorf("standard input is unavailable")
```
maybe so... the fact that we called it stdinInUse in the resource builder is telling

liggitt replied:
or, how about this... we make it `StdinUnavailableMessage string`, the setter can indicate the reason, and we use it here like `standard input is unavailable: %s`

resource builder can set the message to "used by stdin manifest reader" or something

liggitt replied:
lgtm

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -213,6 +213,33 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+	case api.NeverExecInteractiveMode:
+		shouldBeInteractive = false
+	case api.IfAvailableExecInteractiveMode:
+		shouldBeInteractive = isTerminalFunc(int(os.Stdin.Fd())) && !config.StdinUnavailable
```
swap this order to do the cheap boolean check before the `isTerminalFunc` call

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -213,6 +213,33 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
+		}
+		shouldBeInteractive = true
+	default:
+		return false, fmt.Errorf("unknown ExecInteractiveMode: %q", config.InteractiveMode)
```
nit: use the external API field name (`interactiveMode`)

liggitt replied:
also, should this be caught earlier in validation in `staging/src/k8s.io/client-go/tools/clientcmd/validation.go#validateAuthInfo`?

ankeesler replied:
this should be caught in https://github.com/kubernetes/kubernetes/blob/47e248f6fa40cad48ebc30340b6f2be339d25c27/staging/src/k8s.io/client-go/tools/clientcmd/validation.go#L311-L319

unless people are building their own `ExecConfig's` in code? i dunno how much this happens/how much we care. i seem to remember seeing only a handful of grep.app results that showed people doing this

if we did remove this validation, i'm not sure what we would do in the `default` case - seems like we need to return some sort of error if we get into that state (even tho we may never get into it)

i updated the error message in b8e58d6137b

liggitt replied:
oh, I'm fine keeping the check here as well, just hoped it would be caught way earlier in kubeconfig validation as well

ankeesler commented on staging/src/k8s.io/cli-runtime/pkg/resource/client.go:
```diff
@@ -56,3 +56,14 @@ func (clientConfigFn ClientConfigFunc) unstructuredClientForGroupVersion(gv sche
+		cfg, err := clientConfigFn()
+		if stdinUnavailable && cfg != nil && cfg.ExecProvider != nil {
+			cfg.ExecProvider.StdinUnavailable = stdinUnavailable
+			cfg.ExecProvider.StdinUnavailableMessage = "used by cli-runtime stdin resource manifest reader"
```
47e248f6fa40cad48ebc30340b6f2be339d25c27

liggitt commented:
I'd like to see a test added to `test/cmd/authentication.sh` that exercises kubectl commands like `kubectl ... -f -` with an exec plugin with `interactiveMode: Always` to make sure we don't have ordering issues where the unavailability of stdin is registered too late

also, include calls to the kubectl commands that read stdin specially (set env, delete --raw, create --raw, update --raw)

enj commented on staging/src/k8s.io/client-go/tools/clientcmd/api/types.go:
```diff
@@ -245,6 +245,33 @@ type ExecConfig struct {
+	// cannot successfully run this exec plugin because it needs to use standard input and
+	// StdinUnavailable is true. For example, a process that is already using standard input to
+	// read user instructions might set this to "used by my-program to read user instructions".
+	// +k8s:conversion-gen=false
+	StdinUnavailableMessage string
```
?

ankeesler committed:
exec credential provider: InteractiveMode support

The value here is that the exec plugin author can use the kubeconfig to assert
how standard input is treated with respect to the exec plugin, e.g.,
- an exec plugin author can ensure that kubectl fails if it cannot provide
  standard input to an exec plugin that needs it (Always)
- an exec plugin author can ensure that an client-go process will still call an
  exec plugin that prefers standard input even if standard input is not
  available (IfAvailable)

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler committed:
kubectl: don't fight for stdin with exec plugins

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler committed:
test/cmd: kubectl and exec plugins don't fight for stdin

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

liggitt commented:
/lgtm
/approve

ankeesler commented:
Looking into test failure. It might be flake https://github.com/kubernetes/kubernetes/issues/102842#issuecomment-861237027

/hold

ankeesler commented:
pull-kubernetes-integration was an actual test failure, fixed by https://github.com/ankeesler/kubernetes/blob/cd83d89ac94c5b61fdd38840098e7223e5af0d34/test/integration/client/exec_test.go#L693
/hold cancel

pull-kubernetes-unit failure looks like https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=TestMain#bf11e79583121ff73ec4
/test pull-kubernetes-unit

liggitt commented:
/lgtm
